### PR TITLE
fix(daemon): recognize linked prototype page delivery

### PR DESCRIPTION
**Pull request by Siri-Ray**
Related work item: [OPEND-2887](https://plane.powerformer.net/open-design/projects/49832a02-3158-4faf-bf2f-d0e39c40c7e6/issues/b50881bc-e83b-45cf-9079-51908835b163/)

## Why

This follows a user diagnostic bundle where a prototype child page was written successfully, but the OD Next task ended blocked after the agent omitted Runtime State. The host completion fallback only accepted changes to the canonical entry. Writing a page already linked from an unchanged homepage therefore failed delivery validation. Without an explicit entry, adding a second HTML file also made entry inference ambiguous.

The historical bundle lacks entry metadata, so it cannot establish which validation branch ran on that machine. Both branches are reproduced against current main.

- Validate changed prototype pages reachable through local HTML navigation from the canonical entry. Preserve per-run touch evidence and existing coordinator route constraints.
- Retain an unambiguous pre-run entry after verified linked-page delivery, without replacing an explicit selection. Subsequent edits retain the same homepage identity.
- Check syntax on the changed child page and log the bounded delivery-rejection reason.

Traversal reads static local anchors only, with limits of 64 pages, 512 KiB per page and 8 MiB total. Unrelated files, external links and absent write evidence do not qualify. Prompt contracts, automatic repair policy and network-disconnect handling are unchanged. Adjacent #7931 and #7942 concern status/error presentation and finalization policy.

## What users will see

Adding or editing a linked prototype page can complete the OD Next task even when the agent omits Runtime State, provided the existing completion-inference rules allow it. The homepage remains the project entry. A separate confirmation turn that writes nothing cannot borrow the previous task's delivery evidence.

The existing conversation reload and `od run info --json` surfaces consume the same corrected daemon result.

## Surface area

- [ ] **UI** — new page / dialog / panel / menu item / setting / empty state
- [ ] **Keyboard shortcut** — new or changed
- [ ] **CLI / env var** — new subcommand, flag or environment variable
- [ ] **API / contract** — new endpoint, SSE event or changed DTO shape
- [ ] **Extension point** — skills, design systems, templates, craft or protocol
- [ ] **i18n keys** — new translation keys
- [ ] **New top-level dependency** — root package.json dependency
- [x] **Default behavior change** — linked prototype-page delivery and retention of an inferred entry
- [ ] **None** — internal refactor, docs, tests or translation only

No UI component, CLI flag or shared response shape changes; both existing surfaces are exercised by the integration test. Cheerio is already a daemon dependency.

## Screenshots

Not applicable: no UI layout or entry point changes. Tests verify persisted conversation delivery and the CLI response.

## Bug fix verification

- [Validator regressions](https://github.com/Siri-Ray/open-design/blob/50550f8f6407482a48b5abd83c7a9af15b3579dc/apps/daemon/tests/run-deliverable-validation.test.ts#L235) and [real-server regressions](https://github.com/Siri-Ray/open-design/blob/50550f8f6407482a48b5abd83c7a9af15b3579dc/apps/daemon/tests/od-next-automatic-simple-server.test.ts#L2216): **red on main `f197844e27`**, green with the fix. The original four cases failed on delivery/task assertions, not environment setup.
- The real-server test uses the shipped HTTP/CLI seams and a synthetic agent that writes HTML but omits Runtime State. It checks prompt-contract presence, conversation reload, native-session resume, distinct task identities, repeated editing and rejection of a no-write confirmation.
- [Syntax regression](https://github.com/Siri-Ray/open-design/blob/50550f8f6407482a48b5abd83c7a9af15b3579dc/apps/daemon/tests/artifacts/successful-run-deliverable-finalization.test.ts#L123) verifies that an unsafe child-page script produces a warning without changing the homepage or child source.
- Boundary coverage includes disconnected pages, cycles, external/base URLs, encoded names, stale entry metadata, unreliable diffs, other project kinds, oversized pages and external symlinks.

## Validation

- `corepack pnpm guard` — passed.
- `corepack pnpm typecheck` — passed; daemon typecheck repeated after entry retention was added, also passed.
- `corepack pnpm --filter @open-design/daemon exec vitest run tests/run-deliverable-validation.test.ts tests/artifacts/successful-run-deliverable-finalization.test.ts tests/strategies/od-next/coordinator.test.ts tests/od-next-automatic-simple-server.test.ts` — 100 passed, 1 existing skip.
- Final entry-retention HTTP/CLI and linked-page regressions — 14 passed; final validator/finalizer/coordinator suites after the additional boundary case — 68 passed, 1 existing skip.
- `git diff --check` — passed. No live provider calls or release/deployment validation.

### CODE REVIEW

**Comment by lefarcen** [comment]
Queued for QA validation — this PR changes runtime delivery behavior that needs a manual QA pass before merge. Nothing is needed from you right now; we will update this thread once validation is complete.

File: apps/daemon/src/run-deliverable-validation.ts [hunk 1]
```diff
@@ -184,10 +197,15 @@ export async function validateRunDeliverable(
...
   const acceptedKinds = acceptedDeliverableKinds(input.projectMetadata);
+  const isPrototype = projectKind(input.projectMetadata) === 'prototype';
   const declared = safeRelativeFile(input.projectMetadata?.entryFile);
+  const baselineEntry = isPrototype && input.touchedPaths
+    ? safeRelativeFile(input.baselineEntryFile)
+    : null;
   const selected = declared
     ? files.find((file) => filePath(file) === declared) ?? null
-    : inferredEntry(files, acceptedKinds);
+    : inferredEntry(files, acceptedKinds)
```
Comment (nettee) [inline apps/daemon/src/run-deliverable-validation.ts]: **Must fix — preserve the pre-run entry before post-run inference.**

For a metadata-free prototype that starts with only `landing.html` and whose landing page links to a newly written `index.html`, this priority selects the new `index.html` once the run finishes. Because that file is itself touched, `linkedPage` remains unset, so the retention block in `server.ts` never persists `landing.html`. Later runs then infer `index.html`, changing the homepage identity this PR is intended to preserve.

Select a verified `baselineEntry` before `inferredEntry` for metadata-free prototype runs, while keeping the explicit-entry branch authoritative. Please add a regression where `landing.html` links to a newly created `index.html` so the retained entry and subsequent delivery validation stay bound to `landing.html`.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Review by nettee — CHANGES_REQUESTED** [review]
I found one blocking correctness issue in the new baseline-entry retention path; the inline comment includes the concrete failure mode and fix.



<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

File: apps/daemon/tests/run-deliverable-validation.test.ts [hunk 2]
```diff
@@ -231,3 +231,107 @@ describe('run deliverable validation', () => {
...
+      touchedPaths: ['catalog.html'], baselineEntryFile: 'landing.html' };
+    await expect(validateRunDeliverable({ ...input, projectMetadata: { kind: 'prototype' } }))
+      .resolves.toMatchObject({ valid: true, entryFile: 'landing.html', linkedPage: 'catalog.html' });
+    await expect(validateRunDeliverable({ ...input, projectMetadata: { kind: 'prototype', entryFile: 'missing.html' } }))
+      .resolves.toMatchObject({ valid: false, validation: 'entry_missing' });
+    await expect(validateRunDeliverable({ ...input, artifactCount: 0, touchedPaths: [], projectMetadata: { kind: 'prototype' } }))
+      .resolves.toMatchObject({ valid: false, validation: 'no_artifact' });
+    await expect(validateRunDeliverable({ ...baseInput, touchedPaths: input.touchedPaths, projectMetadata: { kind: 'prototype' } }))
+      .resolves.toMatchObject({ valid: false, validation: 'entry_missing' });
+    await expect(validateRunDeliverable({ ...baseInput, baselineEntryFile: input.baselineEntryFile, projectMetadata: { kind: 'prototype' } }))
```
Comment (nettee) [inline apps/daemon/tests/run-deliverable-validation.test.ts:304]: **Must fix — cover and implement the still-missing pre-run-entry priority.**

This amendment only separates the missing-input cases; it still has no case where `landing.html` was the sole pre-run page and the run creates/touches a linked `index.html`. In that scenario `inferredEntry()` selects the newly created `index.html` before the verified `baselineEntryFile`, so the validator treats it as the canonical entry rather than a `linkedPage`. The server's retention branch then cannot persist `landing.html`, and the next run changes the homepage identity—the regression reported in the unresolved prior thread.

Add a regression that supplies `baselineEntryFile: 'landing.html'`, a touched linked `index.html`, and asserts `entryFile === 'landing.html'` plus `linkedPage === 'index.html'` (including the subsequent retained-entry path). Change the metadata-free prototype selection order so a present verified baseline entry is selected before post-run inference; keep an explicit metadata `entryFile` authoritative.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Review by nettee — CHANGES_REQUESTED** [review]
The previous blocking baseline-entry issue remains unresolved: this test-only amendment does not change the post-run selection order. The inline comment identifies the missing regression and implementation change.



<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

File: apps/daemon/src/run-deliverable-validation.ts (hunk 1 above)
Comment (Siri-Ray) [reply apps/daemon/src/run-deliverable-validation.ts]: @nettee This has been fixed in 4c6891c.

Updated run-deliverable-validation.ts to prefer the existing baseline entry before inferredEntry while preserving explicit entry authority. Added validator and real-server regressions covering linked index.html creation, persisted homepage identity, and subsequent delivery; both regressions failed before the repair and passed afterward.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by Siri-Ray** [comment]
**Looper fixer round complete** — 4c6891c

- ✅ Review comment on `apps/daemon/src/run-deliverable-validation.ts` (@nettee) — [thread](https://github.com/nexu-io/open-design/pull/7949#discussion_r3967111597)
  - Updated run-deliverable-validation.ts to prefer the existing baseline entry before inferredEntry while preserving explicit entry authority. Added validator and real-server regressions covering linked index.html creation, persisted homepage identity, and subsequent delivery; both regressions failed before the repair and passed afterward.
- • Review comment on `apps/daemon/tests/run-deliverable-validation.test.ts:304` (@nettee) — [thread](https://github.com/nexu-io/open-design/pull/7949#discussion_r3967180753)
  - Agent did not provide a decision for this thread


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

File: apps/daemon/tests/run-deliverable-validation.test.ts (hunk 2 above)
Comment (Siri-Ray) [reply apps/daemon/tests/run-deliverable-validation.test.ts:304]: @nettee I'm not making a code change for this thread.

Already implemented in 4c6891c: run-deliverable-validation.ts prefers the verified baseline before inference while preserving explicit entry authority. Four focused validator and real-server regressions passed, covering linked index.html creation, persisted homepage identity, and subsequent delivery, so no duplicate change was made.




<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=fixer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Review by nettee — APPROVED** [review]
@Siri-Ray

I reviewed the repair frontier for the prior baseline-entry finding. The validator now selects the verified pre-run prototype entry before post-run inference while keeping an explicit entry authoritative, and the new unit plus real-server scenarios cover a linked newly created `index.html`, entry retention, and a follow-up edit. The targeted test command could not run in this prepared worktree because dependencies are absent, but the published CI is in progress and the changed control flow and fixtures are consistent.

Thanks for closing the loop on this subtle homepage-identity case — the focused regression coverage makes the intent much easier to preserve.



<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by ivy-tinggg** [comment]
@Siri-Ray

Thanks for the contribution. I validated the current head against the linked-prototype delivery and homepage-retention paths. The focused OPEND-2887 real-server regression passed with the extended timeout, the coordinator suite passed 41 tests with one existing skip, daemon typecheck passed, and exact-head CI is green. The earlier baseline-entry blocker is fixed in 4c6891c; one stale unresolved thread remains but the reviewer approved this head. No blocking issue was found. Accepted.

**Review by ivy-tinggg — APPROVED** [review]
